File: Test_automation/Selenium/SwagLabsPOM/src/main/java/pages/InventoryPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

public class InventoryPage
{
    private WebDriver driver;
    private By cartLink = By.cssSelector("[data-test='shopping-cart-link']");
    private By hamburgerIcon = By.id("react-burger-menu-btn");
    private By logoutLink = By.cssSelector("[data-test='logout-sidebar-link']");

    public InventoryPage(WebDriver driver)
    {
        this.driver = driver;
    }

    public boolean isCartLinkVisible()
    {
        return driver.findElement(cartLink).isDisplayed();
    }

    public void logout() throws InterruptedException {
        WebDriverWait webDriverWait = new WebDriverWait(driver, Duration.ofSeconds(5));

        // Explicit wait of 5 seconds or less (if the hamburger icon link becomes clickable before 5th second then it proceeds with the test and doesn't wait)
        webDriverWait.until(ExpectedConditions.elementToBeClickable(hamburgerIcon)).click();

        // Explicit wait of 5 seconds or less (if the logout link becomes clickable before 5th second then it proceeds with the test and doesn't wait)
        webDriverWait.until(ExpectedConditions.elementToBeClickable(logoutLink)).click();
    }
}

File: Test_automation/Playwright/SwagLabsPOM/pages/InventoryPage.js
exports.InventoryPage = class InventoryPage
{
    constructor(page)
    {
        this.page = page;

        this.cartLink = page.locator('[data-test="shopping-cart-link"]');
        this.hamburgerIcon = page.locator('#react-burger-menu-btn');
        this.logoutLink = page.locator('[data-test="logout-sidebar-link"]');
    }

    // Thanks to the Playwright Auto-waiting feature, there is no need for explicit waits
    async logout()
    {
        await this.hamburgerIcon.click();
        await this.logoutLink.click();
    }
}